File: solutions/scala/bank-account/1/src/main/scala/BankAccount.scala
trait BankAccount {

  def closeAccount(): Unit

  def getBalance: Option[Int]

  def incrementBalance(increment: Int): Option[Int]
}

object Bank {
  def openAccount(): BankAccount = new Account

  private class Account extends BankAccount {
    private var balance: Option[Int] = Some(0)

    override def closeAccount(): Unit = synchronized {
      balance = None
    }

    override def getBalance: Option[Int] = synchronized {
      balance
    }

    override def incrementBalance(increment: Int): Option[Int] = synchronized {
      balance = balance.map(_ + increment)
      balance
    }
  }
}

File: solutions/kotlin/bank-account/1/src/main/kotlin/BankAccount.kt
class BankAccount {
    private var currentBalance: Long? = 0

    val balance: Long
        get() =
            synchronized(this) {
                currentBalance ?: throw closedAccount()
            }

    fun adjustBalance(amount: Long) {
        synchronized(this) {
            currentBalance = balance + amount
        }
    }

    fun close() {
        synchronized(this) {
            currentBalance = null
        }
    }

    private fun closedAccount() = IllegalStateException("Account is closed.")
}
